### ISSUE TITLE
Feature Request: Bundle Sources in Maven SDK Distribution

### ISSUE THREAD

**Issue by ParthPadg** [firebase/firebase-android-sdk#6]
## What feature would you like to see?

First off, I think it's awesome that the Firebase SDK is being open-sourced. However, as an Android developer, my primary use-case for open source is being able to explore the sources from inside Android Studio. 
Thus, I think it'd be _even more_ awesome if y'all should ship the sources as part of the Maven bundle that we request through Gradle.

**Comment by bjornick** [comment]
This sounds like a great idea.  I have one question.  Since our SDKs are proguarded, would this still have value if source names (and possibly code) don't have a 1:1 mapping with the classes we produce?

**Comment by ParthPadg** [comment]
Well, my main thought on this is: don't minify the SDK! The cat's already out of the bag, given that we're discussing this on GH! 😄 

However, since large orgs move slowly, having un-minified sources bundled with minified binaries is fine as a stopgap.

**Comment by bjornick** [comment]
The TL;DR of the next bit is, if I am trying to get buy in to stop proguarding the open source SDKs this week.  If I can, I'll make both changes by the next time we release these SDKs. 

Just as an update, I'm trying to get buy-in to stop proguarding our open source SDKs.  
For context, there are two reasons why we proguard the Firebase SDKs:
 - obfuscation:  Some SDKs want to keep their internals, but obvious that doesn't apply to any of the open source sdks.
 - Size impact: As a part of google, we have a vested interest in the android ecosystem, so we should do things that improve the ecosystem. Since most apps aren't proguarded and a lot of apps use some parts of Firebase, we should at least proguard our SDKs to reduce the download and install size of our SDK code.  This is relevant for our open source SDKs, but I suspect that the net impact on the ecosystem is actually pretty small.  I suspect that most of the top apps do proguard anyway (because in many countries were data is expensive, this has a huge impact on how many install an app).We are trying to get some data to figure out the size impact of not proguarding our SDKs.  I think there is a decent chance that I can make some headway on this by the end of the week.

**Comment by bjornick** [comment]
Good news, we'll stop proguarding our open source SDKs.  Would you still want the source jars there as well? If so, I'll track the removal of proguard in a separate issue.

**Comment by ParthPadg** [comment]
I'll admit a bit of ignorance RE difference between source JARs versus a non-minified binary...does one have javadocs + comments out-of-the-box? 
Because then I would prefer that option 😄

**Comment by samtstern** [comment]
@nbransby 

1. The next release will be non-proguarded.
2. As you can see, this repository only contains some of our SDKs (Firestore, Database, Storage, etc).  However many Firebase SDKs like Auth, Analytics, etc are not here yet.  That's what the note is saying.  This is important because some of the SDKs that *are* open source depend on some of the closed-source ones, so you can't yet be open-source all the way down.

**Comment by samtstern** [comment]
An update here: you can now use "Go to definition" in Android Studio to see source code for all of our open-source SDKs:

![image](https://user-images.githubusercontent.com/8466666/48502166-bc7a4900-e7f3-11e8-8119-693f4226a040.png)

However you'll notice that this Java view is produced by Android Studio on the fly by decompiling the `.class` bytecode files.

If we were to ship source JARs, Android Studio (and other tools) would have access to the original `.java` source files not just the `.class` files which can be decompiled.

The question is: is there much more value to this than what we have now?

**Comment by ParthPadg** [comment]
It's an awesome start! But the Android Studio tooling is not as string with decompiled sources as it is with actual sources -- things like _Find Usage Of_ don't work as well. 

It's definitely not a blocker, but the point of shipping un-minified is to aid in developer understanding...so this seems like the logical next step.

**Comment by matejdro** [comment]
I my opinion, including original java sources brings a lot of value:

- It enables IDE to display javadocs directly. This saves a lot of trouble of having to go to the reference page every time
- It displays detailed deprecation warnings. If method is deprecated, IDE shows it crossed out, but without original java files we can only see the `@Deprecated` annotation without any explanation. Again, saves the trip to the reference page

**Comment by vkryachko** [comment]
Hey folks, great news.

All releases of our open-source SDKs going forward will contain sources.jar. The first one to come out is firestore and firestore-ktx in the coming days.
